File: rust/src/bin/request_lets_encrypt_certificate.rs
use dnsimple::dnsimple::new_client;
use dnsimple_rust_sample_code::token_from_file;
use std::env;

fn main() {
    let token = token_from_file();
    let args: Vec<String> = env::args().collect();

    request_lets_encrypt_certificate(
        &*token,
        &args[1],
        args.last().unwrap().parse::<u64>().unwrap(),
    );
}

pub fn request_lets_encrypt_certificate(token: &str, domain_name: &str, certificate_id: u64) {
    // Construct a client instance
    //
    // If you want to connect to production set the sandbox argument to false
    let client = new_client(true, token.to_string());

    // All calls to client pass through a service. In this case, `client.identity()` is the identity
    // service.
    //
    // `client.identity().whoami` is the method for retrieving the account details for your
    // current credentials via the DNSimple API.
    let whoami = client.identity().whoami().unwrap().data.unwrap();

    // The response object returned by any API method includes a `data` attribute. Underneath
    // that attribute is an attribute for each data object returned. In this case, `account` provides
    // access to the resulting account object.
    //
    // Here the account ID is extracted for use in the call to cancel domain transfer.
    let account_id = whoami.account.unwrap().id;

    let response = client.certificates().issue_letsencrypt_certificate(
        account_id,
        domain_name,
        certificate_id,
    );

    println!(
        "Certificate for domain {} requested: {:?}",
        domain_name,
        response.unwrap().data.unwrap()
    );
}

File: ruby/request_letsencrypt_certificate.rb
#!/usr/bin/env ruby

# Usage: ENV["CERTIFICATE_DOMAIN_ID"]=example.com ruby ./request_letsencrypt_certificate.rb

require 'pp'
require 'dnsimple'
require_relative 'token'

ENV["CERTIFICATE_DOMAIN_ID"] or
    abort("set CERTIFICATE_DOMAIN_ID to the ID of the domain to request the certificate for")

# Construct a client instance.
#
# If you want to connect to production, omit the `base_url` option.
client = Dnsimple::Client.new(base_url: "https://api.sandbox.dnsimple.com", access_token: TOKEN)

# All calls to client pass through a service. In this case, `client.identity` is the identity service.
#
# Dnsimple::Client::Identity#whoami is the method for retrieving the account details for your
# current credentials via the DNSimple API.
response = client.identity.whoami

# The response object returned by any API method includes a `data` attribute. Underneath that
# attribute is an attribute for each data object returned. In this case, `#account` provides
# access to the resulting account object.
#
# Here the account ID is extracted for use in the call to create contact.
account_id = response.data.account.id

# Dnsimple::Client::Certificates#purchase_letsencrypt_certificate is the method to create a certificate order.
response = client.certificates.purchase_letsencrypt_certificate(account_id, ENV["CERTIFICATE_DOMAIN_ID"], { name: ENV["CERTIFICATE_NAME"] || "www" })

# Pretty-print the entire response object so you can see what is inside.
pp response

# Dnsimple::Client::Certificates#create_contact is the method to submit the order for issuace.
response = client.certificates.issue_letsencrypt_certificate(account_id, ENV["CERTIFICATE_DOMAIN_ID"], response.data.id)

# Pretty-print the entire response object so you can see what is inside.
pp response
